lineax/_solver/gmres.py

```diff
@@ -257,7 +257,7 @@ def main_gmres(y):
                 lambda x: jnp.pad(x[..., None], ((0, 0),) * x.ndim + ((0, restart),)),
                 r_normalised,
             )
-            coeff_mat_init = jnp.eye(restart, restart + 1, dtype=r_norm.dtype)
+            coeff_mat_init = jnp.eye(restart, restart + 1, dtype=r_normalised[0].dtype)
```

Comment:
I think this is wrong. `r_normalised` is an arbitrary pytree:
https://github.com/google/lineax/blob/26133fa1c92ad096c25a77ef5493a49bc8738332/lineax/_solver/gmres.py#L399
so it may not support indexing like this.

I'm also not sure whether these coefficients should ever be complex-valued, or if they should always be real-valued. @packquickly any thoughts?

Reply:
I think it should be complex. It is a Gram–Schmidt process, and the coefficients in GS can be complex. Also, it fixed some tests :)
Should it be something like `jnp.result_type(jtu.tree_flatten(r_normalised)[0])` then? I'm not too proficient with PyTrees

Reply:
More or less, yup: `jnp.result_type(*jtu.tree_leaves(r_normalised))`, I think.